File: game_of_agents/settlement.py
from __future__ import annotations

from collections.abc import Iterable, Mapping
from typing import Any

from game_of_agents.models import SettlementMode


def _purchase_field(purchase: Any, field: str) -> Any:
    if isinstance(purchase, dict):
        return purchase.get(field)
    return getattr(purchase, field, None)


def compute_marketplace_payouts(
    base_scores: Mapping[str, float],
    purchases: Iterable[Any],
    settlement_mode: SettlementMode | str,
) -> dict[str, float]:
    payouts = {agent_id: float(score) for agent_id, score in base_scores.items()}
    mode = settlement_mode.value if isinstance(settlement_mode, SettlementMode) else str(settlement_mode)
    for purchase in purchases:
        buyer = str(_purchase_field(purchase, "buyer_agent_id") or "")
        seller = str(_purchase_field(purchase, "seller_agent_id") or "")
        if not buyer or not seller:
            continue
        transfer = float(base_scores.get(buyer, 0.0)) * (float(_purchase_field(purchase, "price_pct") or 0.0) / 100.0)
        if mode == SettlementMode.NET.value:
            payouts[buyer] = payouts.get(buyer, float(base_scores.get(buyer, 0.0))) - transfer
        payouts[seller] = payouts.get(seller, float(base_scores.get(seller, 0.0))) + transfer
    return payouts

File: dashboard/app/ui/settlement.ts
import type { LeaderboardAgent, Purchase, RunConfig } from "./types";

export type SettlementRow = {
  agentId: string;
  baseScore: number;
  payout: number;
  delta: number;
  status: string;
  bestBotId: string | null;
};

export function formatSignedDelta(value: number): string {
  const rounded = value.toFixed(1);
  return value >= 0 ? `(+${rounded})` : `(${rounded})`;
}

export function displayDelta(value: number): string | null {
  return Math.abs(value) >= 0.05 ? formatSignedDelta(value) : null;
}

export function buildSettlementRows(
  agents: LeaderboardAgent[],
  purchases: Record<string, Purchase> | Purchase[] | undefined,
  settlementMode: RunConfig["settlement_mode"] | undefined,
  finalScores?: Record<string, number> | null,
  payouts?: Record<string, number> | null,
): SettlementRow[] {
  const baseScores = Object.fromEntries(
    agents.map((agent) => [agent.agentId, Number(finalScores?.[agent.agentId] ?? agent.bestRating ?? 0)]),
  );
  const projected = Object.fromEntries(
    agents.map((agent) => [agent.agentId, Number(payouts?.[agent.agentId] ?? baseScores[agent.agentId] ?? 0)]),
  );

  if (!payouts || Object.keys(payouts).length === 0) {
    const purchaseList = Array.isArray(purchases) ? purchases : Object.values(purchases ?? {});
    for (const purchase of purchaseList) {
      const buyer = purchase.buyer_agent_id;
      const seller = purchase.seller_agent_id;
      if (!buyer || !seller) {
        continue;
      }
      const transfer = (baseScores[buyer] ?? 0) * (Number(purchase.price_pct ?? 0) / 100);
      if (settlementMode === "net") {
        projected[buyer] = (projected[buyer] ?? baseScores[buyer] ?? 0) - transfer;
      }
      projected[seller] = (projected[seller] ?? baseScores[seller] ?? 0) + transfer;
    }
  }

  return agents
    .map((agent) => {
      const baseScore = baseScores[agent.agentId] ?? 0;
      const payout = projected[agent.agentId] ?? baseScore;
      return {
        agentId: agent.agentId,
        baseScore,
        payout,
        delta: payout - baseScore,
        status: agent.status,
        bestBotId: agent.bestBotId,
      };
    })
    .sort((left, right) => right.payout - left.payout || right.baseScore - left.baseScore || left.agentId.localeCompare(right.agentId));
}
